htsfile.1
.TH htsfile 1 "9 July 2026" "htslib-1.24" "Bioinformatics tools"
.SH NAME
htsfile \- identify high-throughput sequencing data files
.\"
.\" Copyright (C) 2015, 2017-2018 Genome Research Ltd.
.\"
.\" Author: John Marshall <jm18@sanger.ac.uk>
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a
.\" copy of this software and associated documentation files (the "Software"),
.\" to deal in the Software without restriction, including without limitation
.\" the rights to use, copy, modify, merge, publish, distribute, sublicense,
.\" and/or sell copies of the Software, and to permit persons to whom the
.\" Software is furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
.\" THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
.\" FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
.\" DEALINGS IN THE SOFTWARE.
.\"
.SH SYNOPSIS
.B htsfile
.RB [ -chHv ]
.IR FILE ...
.br
.B htsfile --copy
.RB [ -v ]
.I FILE DESTFILE
.SH DESCRIPTION
The \fBhtsfile\fR utility attempts to identify what kind of high-throughput
sequencing data files the specified files are, and provides minimal viewing
capabilities for some kinds of data file.
.P
It can identify sequencing data files such as SAM, BAM, and CRAM;
variant calling data files such as VCF and BCF;
index files used to index these data files;
and compressed versions of many of them.
.P
For each \fIFILE\fR given, \fBhtsfile\fP prints a description of the file
format determined, using similar keyword conventions to \fBfile\fP(1):
"text" indicates a textual file that can probably be viewed on a terminal;
"data" indicates binary data;
"sequence", "variant calling", and "index" indicate different categories of
data file.
When it can be identified, the name of the particular file format (such as
"BAM" or "VCF") is printed at the start of the description.
.P
When used to view file contents as text, \fBhtsfile\fP can optionally show
only headers or only data records, but has no other filtering capabilities.
Use \fBsamtools\fR or \fBbcftools\fR if you need more extensive viewing or
filtering capabilities.
.P
Alternatively, when \fB--copy\fR is used, \fBhtsfile\fR takes exactly two
arguments and performs a byte-for-byte copy from \fIFILE\fR to \fIDESTFILE\fR.
This is similar to \fBcp\fR(1), but HTSlib's remote file access facilities
are available for both source and destination.
.P
The following options are accepted:
.TP 4n
.BR -c ", " --view
Instead of identifying the specified files, display a textual representation
of their contents on standard output.
.IP
By default, \fB--view\fR refuses to display files in unknown formats.
When \fB--verbose\fR is also given, the raw contents of such files are
displayed, with non-printable characters shown via C-style "\\x" hexadecimal
escape sequences.
.TP
.BR -C ", " --copy
Instead of identifying or displaying the specified files, copy the source
\fIFILE\fR to the destination \fIDESTFILE\fR.
Only \fB--verbose\fR may be used in conjunction with \fB--copy\fR.
.TP
.BR -h ", " --header-only
Display data file headers only.
Implies \fB--view\fR.
.TP
.BR -H ", " --no-header
When viewing files, display data records only.
.TP
.BR -v ", " --verbose
Display additional warnings and diagnostic messages.
Using \fB--verbose\fR repeatedly further raises the verbosity.
.PP
.SH SEE ALSO
.IR bcftools (1),
.IR file (1),
.IR samtools (1)


NEWS
Noteworthy changes in release a.b
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* New environment variable HTS_GCS_HOST, analogous to the existing
  HTS_S3_HOST, to redirect gs:// URLs to a caller-specified endpoint
  (e.g. a local GCS emulator such as fake-gcs-server) using path-style
  addressing instead of the normal BUCKET.storage.googleapis.com
  virtual-hosted-style addressing.

Noteworthy changes in release 1.24 (9th July 2026)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* NEW.  Contribution guidelines, Including sections on complexity, completeness,
  signing and our AI policy.
  (PR #2003)

Updates
-------

* Remove CRAM v4.0 support.  The experimental CRAM v4 code was large, complex
  and insufficiently tested.  With no real move toward v4 adoption we have
  removed the code in order to reduce the likelihood of security issues.
  (PR #2020 and PR #2031.
   Reported by Trail of Bits and Anthropic:
    https://github.com/samtools/htslib/security/advisories/GHSA-6q7c-m967-9v37
   Reported by Team Atlanta:
    https://github.com/samtools/htslib/security/advisories/GHSA-hg3g-v57p-459q
    https://github.com/samtools/htslib/security/advisories/GHSA-rjqv-xg3q-g423)

* Ensure indirect function calls have the correct type.  Some HTSlib interfaces
  have function callbacks that are intended to be generic, so the function
  signatures include a void * for data to be passed in.  While this mostly works
  it is strictly undefined behaviour.  Some new wrapper functions and interfaces
  are added to address this.
  (PR #1994)

* Add wrappers for malloc, realloc with a calloc-like interface.  This helps
  avoid bugs due to integer wrap-around when calculating memory sizes.
  (PR #2006)

* Make faidx work with very long (>4 Gbyte!) lines. Although faidx should
  support very long references, writing one longer than 4Gbases on a single line
  broke it because it used a uint32_t field to store the line length.
  (PR #2008. Fixes samtools/samtools#2331.  Reported by Ying Chen)

* Parallel cram2bam.  Move more of the work into worker threads to speed
  everything up.
  (PR #2015)

* Improve synced reader error checking.  Ensure memory failures are caught and
  that error are propagated to callers.
  (PR #2024)

* Add tbx and bcf multi-region iterators.
  (PR #2030, fixes #1930.  Requested by Adam Novak.
   Replaces the alternative PRs #1997 and #2022)

* Remove references to the non-existent FAI_CACHE.
  (PR #2033, fixes #2032.  Reported by John Marshall)

* The underlying causes of errors when accessing files in S3 buckets are now
  reported more accurately.
  (PR #2036.  Thanks to John Marshall)

* In VCF, improve the "not defined in the header" messages.
  (PR #2007)


Build Changes
-------------

* Add Github actions builds for Linux and Mac OS.  This replaces the Cirrus CI
  tests that were discontinued.
  (PR #2000)

* Ensure PACKAGE_VERSION is set in the Makefile.
  (PR #2038.  See also samtools/samtools#2337)

Bug fixes
---------

* Better enforce QNAME length restrictions in the CRAM reader.
  HTSlib limits the length of QNAMEs (and mate QNAMEs) to 254 characters to
  match the maximum allowed by the BAM format.  To avoid problems with
  handling very long names, enforcement of this limit has been moved to an
  earlier stage in the CRAM decoder.
  (PR #2046.  Reported by Claude and Ada Logics.
   https://github.com/samtools/htslib/security/advisories/GHSA-2jx2-wm7f-rv9m)

* Bounds-check refid when loading a CRAM .crai index.
  (PR #2029.  Thanks to Sidhartha Kumar.
   https://github.com/samtools/htslib/security/advisories/GHSA-28r7-prwc-hf5c)

* Improve checks of record consitency in the BCF reader, and disallow
  excessively large FORMAT entries.
  (PR #2047.  Reported by VulnSeeker Security Research.
   https://github.com/samtools/htslib/security/advisories/GHSA-2wm6-8hgm-7g92)

* Improve validation of faidx index entries.
  (PR #2008.
   https://github.com/samtools/htslib/security/advisories/GHSA-4hjq-r829-8c8v)

* Several fixes co-authored by Team Atlanta.
  - Possible integer overflow when checking input size for the beta codec.
  - Oversized shift in cram_subexp_decode.
  (PR #2025, #2046.  Thanks to Team Atlanta.
   https://github.com/samtools/htslib/security/advisories/GHSA-rjqv-xg3q-g423)

* Fix s3_seek returning wrong offset on cache-hit.
  This caused unreliability reading of S3 data when using the
  multi-region iterator. The bug affects 1.23 and 1.23.1.
  Releases prior to 1.23 were unaffected.
  (PR #2012.  Thanks to Nick Edwards.
   Also fixes #2043 reported by Andrew J. Tock.)

* Protect against uninitialised variable read with X_NOSZ codecs.
  (PR #2023, fixes #2021.  Reported by Jiami Lin)

* Fix a possible crash when using nonsensical embed_ref=2,no_ref CRAM encode
  options.
  (PR #2048)

* Fix a read buffer overflow and improve SQ LN field checking.
  The buffer overflow was in development code only and never in a release.
  (PR #1999. Fixes oss-fuzz issue 499447432)

* Small updates to fix misplaced NULL checks, remove unused variables, and
  ensure the package builds without warnings on newer compilers.
  (PR #1992, PR #1993, PR #2004, PR #2010)

* Remove a signed overflow bug in bgzf_read_small.
  (PR #2013)

* In bgzip and tabix, when handling errors, don't try to access the
  BGZF::errcode field after calling bgzf_close() as it may no longer be valid.
  (PR #2035, #2042)

* Speed up hts_parse_decimal() handling of oversized exponents
  (PR #2045)

* Fix some memory leaks on failing to build or load an index
  (PR #2049)

Documentation updates
---------------------

* Clarify how defaults work when building an index with tabix.
  (PR #2002, addresses #1995. Query made by Dario Beraldi)


Noteworthy changes in release 1.23.1 (18th March 2026)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Bug fixes
---------

* Fix a number of bugs in the CRAM decoder which could result in undefined
  behaviour on invalid inputs (PR #1981, PR #1991):


htsfile.c
/*  htsfile.c -- file identifier and minimal viewer.

    Copyright (C) 2014-2019 Genome Research Ltd.

    Author: John Marshall <jm18@sanger.ac.uk>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.  */

#include <config.h>

#include <ctype.h>
#include <errno.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>

#include "htslib/hfile.h"
#include "htslib/hts.h"
#include "htslib/sam.h"
#include "htslib/vcf.h"

#ifndef EFTYPE
#define EFTYPE ENOEXEC
#endif

enum { identify, view_headers, view_all, copy } mode = identify;
int show_headers = 1;
int verbose = 0;
int status = EXIT_SUCCESS;  /* Exit status from main */

void HTS_FORMAT(HTS_PRINTF_FMT, 1, 2) error(const char *format, ...)
{
    int err = errno;
    va_list args;
    va_start(args, format);
    fflush(stdout);
    fprintf(stderr, "htsfile: ");
    vfprintf(stderr, format, args);
    if (err) fprintf(stderr, ": %s\n", strerror(err));
    else fprintf(stderr, "\n");
    fflush(stderr);
    va_end(args);
    status = EXIT_FAILURE;
}

static void view_sam(samFile *in, const char *filename)
{
    bam1_t *b = NULL;
    sam_hdr_t *hdr = NULL;
    samFile *out = NULL;

    hdr = sam_hdr_read(in);
    if (hdr == NULL) {
        errno = 0; error("reading headers from \"%s\" failed", filename);
        goto clean;
    }

    out = hts_open("-", "w");
    if (out == NULL) { error("reopening standard output failed"); goto clean; }

    if (show_headers) {
        if (sam_hdr_write(out, hdr) != 0) {
            error("writing headers to standard output failed");
            goto clean;
        }
    }

    if (mode == view_all) {
        int ret;

        b = bam_init1();
        if (b == NULL) { error("can't create record"); goto clean; }

        while ((ret = sam_read1(in, hdr, b)) >= 0) {
            if (sam_write1(out, hdr, b) < 0) {
                error("writing to standard output failed");
                goto clean;
            }
        }

        if (ret < -1) { error("reading \"%s\" failed", filename); goto clean; }
    }

 clean:
    sam_hdr_destroy(hdr);
    bam_destroy1(b);
    if (out) hts_close(out);
}

static void view_vcf(vcfFile *in, const char *filename)
{
    bcf1_t *rec = NULL;
    bcf_hdr_t *hdr = NULL;
    vcfFile *out = NULL;

    hdr = bcf_hdr_read(in);
    if (hdr == NULL) {
        errno = 0; error("reading headers from \"%s\" failed", filename);
        goto clean;
    }

    out = hts_open("-", "w");
    if (out == NULL) { error("reopening standard output failed"); goto clean; }

    if (show_headers) {
        if (bcf_hdr_write(out, hdr) != 0) {
            error("writing headers to standard output failed");
            goto clean;
        }
    }

    if (mode == view_all) {
        int ret;

        rec = bcf_init();
        if (rec == NULL) { error("can't create record"); goto clean; }

        while ((ret = bcf_read(in, hdr, rec)) >= 0) {
            if (bcf_write(out, hdr, rec) < 0) {
                error("writing to standard output failed");
                goto clean;
            }
        }

        if (ret < -1) { error("reading \"%s\" failed", filename); goto clean; }
    }

 clean:
    if (hdr) bcf_hdr_destroy(hdr);
    if (rec) bcf_destroy(rec);
    if (out) hts_close(out);
}

static void view_raw(hFILE *fp, const char *filename)
{
    int c, prev;
    for (prev = '\n'; (c = hgetc(fp)) != EOF; prev = c)
        if (isprint(c) || c == '\n' || c == '\t') putchar(c);
        else if (c == '\r') fputs("\\r", stdout);
        else if (c == '\0') fputs("\\0", stdout);
        else printf("\\x%02x", c);

    if (prev != '\n') putchar('\n');

    if (herrno(fp)) {
        errno = herrno(fp);
        error("reading \"%s\" failed", filename);
    }
}

static void copy_raw(const char *srcfilename, const char *destfilename)
{
    hFILE *src = hopen(srcfilename, "r");
    if (src == NULL) {
        error("can't open \"%s\"", srcfilename);
        return;
    }

    size_t bufsize = 1048576;
    char *buffer = malloc(bufsize);
    if (buffer == NULL) {
        error("can't allocate copy buffer");
        hclose_abruptly(src);
        return;
    }

    hFILE *dest = hopen(destfilename, "w");
    if (dest == NULL) {
        error("can't create \"%s\"", destfilename);
        hclose_abruptly(src);
        free(buffer);
        return;
    }

    ssize_t n;
    while ((n = hread(src, buffer, bufsize)) > 0)
        if (hwrite(dest, buffer, n) != n) {
            error("writing to \"%s\" failed", destfilename);
            hclose_abruptly(dest);
            dest = NULL;
            break;
        }

    if (n < 0) {
        error("reading from \"%s\" failed", srcfilename);
        hclose_abruptly(src);
        src = NULL;
    }

    if (dest && hclose(dest) < 0) error("closing \"%s\" failed", destfilename);
    if (src && hclose(src) < 0)   error("closing \"%s\" failed", srcfilename);
    free(buffer);
}

static void usage(FILE *fp, int status)
{
    fprintf(fp,
"Usage: htsfile [-chHv] FILE...\n"
"       htsfile --copy [-v] FILE DESTFILE\n"
"Options:\n"
"  -c, --view         Write textual form of FILEs to standard output\n"
"  -C, --copy         Copy the exact contents of FILE to DESTFILE\n"
"  -h, --header-only  Display only headers in view mode, not records\n"
"  -H, --no-header    Suppress header display in view mode\n"
"  -v, --verbose      Increase verbosity of warnings and diagnostics\n");
    exit(status);
}

int main(int argc, char **argv)
{
    static const struct option options[] = {
        { "copy", no_argument, NULL, 'C' },
        { "header-only", no_argument, NULL, 'h' },
        { "no-header", no_argument, NULL, 'H' },
        { "view", no_argument, NULL, 'c' },
        { "verbose", no_argument, NULL, 'v' },
        { "help", no_argument, NULL, 2 },
        { "version", no_argument, NULL, 1 },
        { NULL, 0, NULL, 0 }
    };

    int c, i;

    status = EXIT_SUCCESS;
    while ((c = getopt_long(argc, argv, "cChHv", options, NULL)) >= 0)
        switch (c) {
        case 'c': mode = view_all; break;
        case 'C': mode = copy; break;
        case 'h': mode = view_headers; show_headers = 1; break;
        case 'H': show_headers = 0; break;
        case 'v': hts_verbose++; verbose++; break;
        case 1:
            printf(
"htsfile (htslib) %s\n"
"Copyright (C) 2026 Genome Research Ltd.\n",
                   hts_version());
            exit(EXIT_SUCCESS);
            break;
        case 2:   usage(stdout, EXIT_SUCCESS); break;
        default:  usage(stderr, EXIT_FAILURE); break;
        }

    if (optind == argc) usage(stderr, EXIT_FAILURE);

    if (mode == copy) {
        if (optind + 2 != argc) usage(stderr, EXIT_FAILURE);
        copy_raw(argv[optind], argv[optind + 1]);
        return status;
    }

    for (i = optind; i < argc; i++) {
        hFILE *fp = hopen(argv[i], "r");
        if (fp == NULL) {
            error("can't open \"%s\"", argv[i]);
            continue;
        }

        if (mode == identify) {
            htsFormat fmt;
            if (hts_detect_format2(fp, argv[i], &fmt) < 0) {
                error("detecting \"%s\" format failed", argv[i]);
                hclose_abruptly(fp);
                continue;
            }

            char *description = hts_format_description(&fmt);
            printf("%s:\t%s\n", argv[i], description);
            free(description);
        }
        else {
            htsFile *hts = hts_hopen(fp, argv[i], "r");
            if (hts) {
                switch (hts_get_format(hts)->category) {
                case sequence_data:
                    view_sam(hts, argv[i]);
                    break;
                case variant_data:
                    view_vcf(hts, argv[i]);
                    break;
                default:
                    if (verbose)
                        view_raw(fp, argv[i]);
                    else {
                        errno = 0;
                        error("can't view \"%s\": unknown format", argv[i]);
                    }
                    break;
                }

                if (hts_close(hts) < 0) error("closing \"%s\" failed", argv[i]);
                fp = NULL;
            }
            else if ((errno == EFTYPE || errno == ENOEXEC) && verbose)
                view_raw(fp, argv[i]);
            else
                error("can't view \"%s\"", argv[i]);
        }

        if (fp && hclose(fp) < 0) error("closing \"%s\" failed", argv[i]);
    }

    if (fclose(stdout) != 0 && errno != EBADF)
        error("closing standard output failed");

    return status;
}
